networks/rpc/http_datadog.go

```diff
@@ -0,0 +1,145 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	httptrace "gopkg.in/DataDog/dd-trace-go.v1/contrib/net/http"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace"
+	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
+	"net/http"
+	"os"
+	"strings"
+)
+
+type DatadogTracer struct {
+	Tags           []string
+	Service        string
+	KlaytnResponse bool
+}
+
+func newDatadogTracer() *DatadogTracer {
+	ddTraceEnabled := os.Getenv("DD_TRACE_ENABLED")
+	if strings.ToLower(ddTraceEnabled) != "true" {
+		return nil
+	}
+
+	tracer.Start()
+
+	tags := strings.Split(os.Getenv("DD_TRACE_HEADER_TAGS"), ",")
+	service := os.Getenv("DD_SERVICE")
+	klaytnResponse := strings.ToLower(os.Getenv("DD_KLAYTN_RPC_RESPONSE")) == "true"
+
+	return &DatadogTracer{tags, service, klaytnResponse}
+}
+
+func newDatadogHTTPHandler(ddTracer *DatadogTracer, handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		defer func() {
+			if err := recover(); err != nil {
+				logger.ErrorWithStack("Datadog http handler panic", "err", err)
+			}
+		}()
+
+		reqMethod := ""
+		reqParam := ""
+
+		// parse RPC requests
+		reqs, isBatch, err := getRPCRequests(r)
+
+		if err != nil || len(reqs) < 1 {
+			// The error will be handled in `handler.ServeHTTP()` and printed with `printRPCErrorLog()`
+			logger.Debug("failed to parse RPC request", "err", err, "len(reqs)", len(reqs))
+		} else {
+			reqMethod = reqs[0].method
+			if isBatch {
+				reqMethod += "_batch"
+			}
+			encoded, _ := json.Marshal(reqs[0].params)
+			reqParam = string(encoded)
+		}
+
+		// datadog transaction name contains the first API method of the request
+		resource := r.Method + " " + r.URL.String() + " " + reqMethod
+
+		// duplicate writer
+		dupW := &dupWriter{
+			ResponseWriter: w,
+			body:           bytes.NewBufferString(""),
+		}
+
+		spanOpts := []ddtrace.StartSpanOption{
+			tracer.Tag("request.method", reqMethod),
+			tracer.Tag("request.params", reqParam),
+		}
+
+		for _, header := range ddTracer.Tags {
```

Comment:
It seems that `ddTracer.Tags` has always the same value during the process lifecycle since the value is parsed from `DD_TRACE_HEADER_TAGS` os env.. Am I right?

Reply:
Yes. `ddTracer.Tags` has always the same value during the process lifecycle.

Reply:
Then, we better move this logic to `newDatadogTracer`. This handler function will be executed on each API call.

Reply:
ddTracer.Tags has always same value. But, this is only used as the key of the tag, and the value is set differently for each request.

```golang
if header == "remote_addr" {
	tag = tracer.Tag(key, r.RemoteAddr)
} else {
	tag = tracer.Tag(key, r.Header.Get(header))
}
```

The value of tag is different for every request, depending on the headers value.

Reply:
I agree. However, the following code can be moved
```
			header = strings.TrimSpace(header)
			header = strings.ToLower(header)
			key := fmt.Sprintf("http.%s", header)
```

Reply:
I agree with @aidan-kwon opinion. This loop is executed every request even though it returns the same result like constants. The lines can be moved in `newDatadogTracer` function.

Reply:
@aidan-kwon @odg0318 good point. I hadn't thought about that part. 
I'll change it.